### PR TITLE
Add a full keyboard device

### PR DESCRIPTION
**Pull request by jdeeny**
Seems to work ok, but I haven't made a robust test in Octo.

### CODE REVIEW

**Comment by jdeeny** [comment]
Here is the demo program reworked to use the keyboard device. XOChip required because it uses non-incrementing save/load.

```# Chip8 is a virtual machine designed in 1977 for programming video games.
# Octo is a high level assembler, disassembler and simulator for Chip8.
# Click 'Run' and then press ASWD to move the sprite around the screen.
# Click the Octo logo for source, documentation and examples.

:alias px v3
:alias py v4


:macro if_keycode code {
	i := 0x1F2
	v0 := code
	save v0 - v0
	load v0 - v0
	if v0 != 0 
}

: main
	i := 0x1F0		# Map keyboard
	v0 := 0x86
	save v0 - v0
	
	px := random 0b0011111
	py := random 0b0001111
	i  := person
	sprite px py 8

	loop

		# erase the player, update its position and then redraw:
		sprite px py 8

		if_keycode 87 then py += -1 # keyboard W
		if_keycode 83 then py +=  1 # keyboard S
		if_keycode 65 then px += -1 # keyboard A
		if_keycode 68 then px +=  1 # keyboard D
		if_keycode 16 then buzz # Keyboard shift
		i := person
		sprite px py 8

		# lock the framerate of this program via the delay timer:
		loop
			vf := delay
			if vf != 0 then
		again
		vf := 3
		delay := vf
	again

: buzz vF := 3 buzzer := vF ;

: person
	0x70 0x70 0x20 0x70 0xA8 0x20 0x50 0x50
```

**Comment by jdeeny** [comment]
This one prints out letters that are typed. Makes it clear there is a conflict with the octo debug keys.

```# Glyph Table
: glyph-table
0x04 0x40 0xA0 0xA0 0xA0 0x40 0x00                   # GLYPH_0 "0" 4px
0x02 0x80 0x80 0x80 0x80 0x80 0x00                   # GLYPH_1 "1" 2px
0x04 0xE0 0x20 0xE0 0x80 0xE0 0x00                   # GLYPH_2 "2" 4px
0x04 0xE0 0x20 0xE0 0x20 0xE0 0x00                   # GLYPH_3 "3" 4px
0x04 0xA0 0xA0 0xE0 0x20 0x20 0x00                   # GLYPH_4 "4" 4px
0x04 0xE0 0x80 0xE0 0x20 0xE0 0x00                   # GLYPH_5 "5" 4px
0x04 0x80 0x80 0xE0 0xA0 0xE0 0x00                   # GLYPH_6 "6" 4px
0x04 0xE0 0x20 0x20 0x20 0x20 0x00                   # GLYPH_7 "7" 4px
0x04 0xE0 0xA0 0xE0 0xA0 0xE0 0x00                   # GLYPH_8 "8" 4px
0x04 0xE0 0xA0 0xE0 0x20 0x20 0x00                   # GLYPH_9 "9" 4px
0x04 0x40 0xA0 0xE0 0xA0 0xA0 0x00                   # GLYPH_A "A" 4px
0x04 0xC0 0xA0 0xC0 0xA0 0xC0 0x00                   # GLYPH_B "B" 4px
0x03 0xC0 0x80 0x80 0x80 0xC0 0x00                   # GLYPH_C "C" 3px
0x04 0xC0 0xA0 0xA0 0xA0 0xC0 0x00                   # GLYPH_D "D" 4px
0x03 0xC0 0x80 0xC0 0x80 0xC0 0x00                   # GLYPH_E "E" 3px
0x03 0xC0 0x80 0xC0 0x80 0x80 0x00                   # GLYPH_F "F" 3px
0x05 0xE0 0x80 0xA0 0x90 0xF0 0x10                   # GLYPH_G "G" 5px
0x04 0xA0 0xA0 0xE0 0xA0 0xA0 0x00                   # GLYPH_H "H" 4px
0x02 0x80 0x80 0x80 0x80 0x80 0x00                   # GLYPH_I "I" 2px
0x04 0x20 0x20 0x20 0xA0 0xE0 0x00                   # GLYPH_J "J" 4px
0x04 0xA0 0xC0 0x80 0xC0 0xA0 0x00                   # GLYPH_K "K" 4px
0x03 0x80 0x80 0x80 0x80 0xC0 0x00                   # GLYPH_L "L" 3px
0x06 0x88 0xD8 0xA8 0x88 0x88 0x00                   # GLYPH_M "M" 6px
0x05 0x90 0x90 0xD0 0xB0 0x90 0x00                   # GLYPH_N "N" 5px
0x04 0xE0 0xA0 0xA0 0xA0 0xE0 0x00                   # GLYPH_O "O" 4px
0x04 0xE0 0xA0 0xE0 0x80 0x80 0x00                   # GLYPH_P "P" 4px
0x04 0xE0 0xA0 0xA0 0xA0 0xE0 0x30                   # GLYPH_Q "Q" 4px
0x04 0xE0 0xA0 0xE0 0xC0 0xA0 0x00                   # GLYPH_R "R" 4px
0x04 0xE0 0x80 0xE0 0x20 0xE0 0x00                   # GLYPH_S "S" 4px
0x04 0xE0 0x40 0x40 0x40 0x40 0x00                   # GLYPH_T "T" 4px
0x04 0xA0 0xA0 0xA0 0xA0 0xE0 0x00                   # GLYPH_U "U" 4px
0x04 0xA0 0xA0 0xA0 0xA0 0x40 0x00                   # GLYPH_V "V" 4px
0x06 0x88 0x88 0x88 0xA8 0x50 0x00                   # GLYPH_W "W" 6px
0x04 0xA0 0xA0 0x40 0xA0 0xA0 0x00                   # GLYPH_X "X" 4px
0x04 0xA0 0xA0 0x40 0x40 0x40 0x00                   # GLYPH_Y "Y" 4px
0x04 0xE0 0x20 0x40 0x80 0xE0 0x00                   # GLYPH_Z "Z" 4px
0x04 0x00 0x00 0x60 0xA0 0x60 0x00                   # GLYPH_a "a" 4px
0x04 0x80 0x80 0xC0 0xA0 0xC0 0x00                   # GLYPH_b "b" 4px
0x04 0x00 0x60 0x80 0x80 0x60 0x00                   # GLYPH_c "c" 4px
0x04 0x20 0x20 0x60 0xA0 0x60 0x00                   # GLYPH_d "d" 4px
0x04 0x00 0x40 0xA0 0xC0 0x60 0x00                   # GLYPH_e "e" 4px
0x03 0x40 0x80 0xC0 0x80 0x80 0x00                   # GLYPH_f "f" 3px
0x04 0x00 0x40 0xA0 0x60 0x20 0xC0                   # GLYPH_g "g" 4px
0x04 0x00 0x80 0x80 0xE0 0xA0 0x00                   # GLYPH_h "h" 4px
0x02 0x00 0x80 0x00 0x80 0x80 0x00                   # GLYPH_i "i" 2px
0x03 0x00 0x40 0x00 0x40 0x40 0x80                   # GLYPH_j "j" 3px
0x04 0x80 0x80 0xA0 0xC0 0xA0 0x00                   # GLYPH_k "k" 4px
0x02 0x80 0x80 0x80 0x80 0x80 0x00                   # GLYPH_l "l" 2px
0x06 0x00 0x00 0x50 0xA8 0xA8 0x00                   # GLYPH_m "m" 6px
0x04 0x00 0x00 0x40 0xA0 0xA0 0x00                   # GLYPH_n "n" 4px
0x04 0x00 0x00 0x40 0xA0 0x40 0x00                   # GLYPH_o "o" 4px
0x04 0x00 0x00 0xC0 0xA0 0xC0 0x80                   # GLYPH_p "p" 4px
0x04 0x00 0x00 0x40 0xA0 0x60 0x20                   # GLYPH_q "q" 4px
0x03 0x00 0x00 0xC0 0x80 0x80 0x00                   # GLYPH_r "r" 3px
0x03 0x00 0x40 0x80 0x40 0x80 0x00                   # GLYPH_s "s" 3px
0x03 0x00 0x80 0xC0 0x80 0x40 0x00                   # GLYPH_t "t" 3px
0x04 0x00 0x00 0xA0 0xA0 0xE0 0x00                   # GLYPH_u "u" 4px
0x04 0x00 0x00 0xA0 0xA0 0x40 0x00                   # GLYPH_v "v" 4px
0x06 0x00 0x00 0x88 0xA8 0x50 0x00                   # GLYPH_w "w" 6px
0x04 0x00 0x00 0xA0 0x40 0xA0 0x00                   # GLYPH_x "x" 4px
0x04 0x00 0x00 0xA0 0xE0 0x20 0xC0                   # GLYPH_y "y" 4px
0x04 0x00 0xC0 0x40 0x80 0xC0 0x00                   # GLYPH_z "z" 4px
0x07 0x9C 0x84 0xE4 0xA4 0xE4 0x00                   # GLYPH_67 "67" 7px
0x06 0xD8 0x90 0x90 0x90 0xD8 0x00                   # GLYPH_CC "CC" 6px
0x08 0xEE 0x88 0xEE 0x82 0xEE 0x00                   # GLYPH_ES "ES" 8px
0x02 0x80 0x80 0x80 0x00 0x80 0x00                   # GLYPH_! "!" 2px
0x02 0x00 0x00 0x00 0x00 0x80 0x00                   # GLYPH_. "." 2px

# Glyph Numbers
:const GLYPH_0             0x00 
:const GLYPH_1             0x01 
:const GLYPH_2             0x02 
:const GLYPH_3             0x03 
:const GLYPH_4             0x04 
:const GLYPH_5             0x05 
:const GLYPH_6             0x06 
:const GLYPH_7             0x07 
:const GLYPH_8             0x08 
:const GLYPH_9             0x09 
:const GLYPH_A             0x0A 
:const GLYPH_B             0x0B 
:const GLYPH_C             0x0C 
:const GLYPH_D             0x0D 
:const GLYPH_E             0x0E 
:const GLYPH_F             0x0F 
:const GLYPH_G             0x10 
:const GLYPH_H             0x11 
:const GLYPH_I             0x12 
:const GLYPH_J             0x13 
:const GLYPH_K             0x14 
:const GLYPH_L             0x15 
:const GLYPH_M             0x16 
:const GLYPH_N             0x17 
:const GLYPH_O             0x18 
:const GLYPH_P             0x19 
:const GLYPH_Q             0x1A 
:const GLYPH_R             0x1B 
:const GLYPH_S             0x1C 
:const GLYPH_T             0x1D 
:const GLYPH_U             0x1E 
:const GLYPH_V             0x1F 
:const GLYPH_W             0x20 
:const GLYPH_X             0x21 
:const GLYPH_Y             0x22 
:const GLYPH_Z             0x23 
:const GLYPH_a             0x24 
:const GLYPH_b             0x25 
:const GLYPH_c             0x26 
:const GLYPH_d             0x27 
:const GLYPH_e             0x28 
:const GLYPH_f             0x29 
:const GLYPH_g             0x2A 
:const GLYPH_h             0x2B 
:const GLYPH_i             0x2C 
:const GLYPH_j             0x2D 
:const GLYPH_k             0x2E 
:const GLYPH_l             0x2F 
:const GLYPH_m             0x30 
:const GLYPH_n             0x31 
:const GLYPH_o             0x32 
:const GLYPH_p             0x33 
:const GLYPH_q             0x34 
:const GLYPH_r             0x35 
:const GLYPH_s             0x36 
:const GLYPH_t             0x37 
:const GLYPH_u             0x38 
:const GLYPH_v             0x39 
:const GLYPH_w             0x3A 
:const GLYPH_x             0x3B 
:const GLYPH_y             0x3C 
:const GLYPH_z             0x3D 
:const GLYPH_67            0x3E 
:const GLYPH_CC            0x3F 
:const GLYPH_ES            0x40 
:const GLYPH_!             0x41 
:const GLYPH_.             0x42 


:macro glyph-index idx-reg {
  i := long glyph-table
  i += idx-reg
  i += idx-reg
  i += idx-reg
  i += idx-reg
  i += idx-reg
  i += idx-reg
  i += idx-reg
}


:macro text-sprite-glyph x-reg y-reg glyph-reg height-px {
  glyph-index glyph-reg
  load v0
  sprite x-reg y-reg height-px
  x-reg += v0
}

:macro if_keycode code {
	i := 0x1F2
	v0 := code
	save v0 - v0
	load v0 - v0
	if v0 != 0 
}

: main
  i := 0x1F0		# Map keyboard
	v0 := 0x86
	save v0 - v0

  v3 := 0
  v4 := 0

: mainloop
  v1 := 65
  v2 := GLYPH_A

  loop
    if_keycode v1 then draw_glyph
    : debounce
    if_keycode v1 then jump debounce
    v1 += 1
    v2 += 1
    while v1 != 90
  again
  jump mainloop

: draw_glyph
  text-sprite-glyph v3 v4 v2 6
  vF := 0xC0
  vF &= v3
  if vF == 0 then return
  v3 := 0
  v4 += 7
  ;
```

File: js/devices/keyboard.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+//
+// To use:
+//   1. Write a keycode to byte 0
+//   2. Read byte 0. A result of 0 indicates that key is not pressed. A result of 1 indicates that key is pressed.
+//
+// See https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode for information on keyCode values.
+
+// TODO: Should remove listeners on teardown
+
+var keyboard_keys = {};
```
Comment (jdeeny) [inline js/devices/keyboard.js:19]: How can this be handled so it doesn't use a global?

**Comment by vladimir-menshakov-red** [comment]
I think we can assume XO chip since we adding memory IO. It's not used in the beginning of boot sequence allowing old hardware running it. If boot sequence and mapper initialisation has passed, we can safely assume it's XO

**Comment by whoozle** [comment]
I think we can assume XO chip since we adding memory IO. It's not used in the beginning of boot sequence allowing old hardware running it. If boot sequence and mapper initialisation has passed, we can safely assume it's XO

File: js/devices/keyboard.js (hunk 1 above)
Comment (whoozle) [reply js/devices/keyboard.js:19]: you can pass emulator instance to ctor, I think, I'll fix it when I started reworking initialisation/teardown

**Pull request merged**